File: test_idecon_connection.py
#!/usr/bin/env python3
"""
Test Connessione Bilancia IDECON
IP: 172.16.224.210 | Port: 50000
Protocollo: STX + [comando] + ETX (byte values 0x02 e 0x03)
"""

import socket
import time
import sys

# Configurazione
IDEON_IP = '172.16.224.210'
IDEON_PORT = 50000
TIMEOUT = 5.0

def create_idecon_command(command: str) -> bytes:
    """Costruisce comando protocollo IDECON (STX + cmd + ETX)"""
    return b'\x02' + command.encode('ascii') + b'\x03'

def parse_response(data: bytes) -> str:
    """Parsa risposta rimuovendo STX/ETX e convertendo in stringa"""
    if len(data) < 2:
        return data.decode('ascii', errors='replace')
    
    # Rimuovi STX (0x02) e ETX (0x03)
    if data[0] == 0x02:
        data = data[1:]
    if data[-1] == 0x03:
        data = data[:-1]
    
    return data.decode('ascii', errors='replace')

def test_connection():
    """Test completo connessione bilancia"""
    print(f"{'='*60}")
    print(f"TEST CONNESSIONE BILANCIA IDECON")
    print(f"{'='*60}")
    print(f"Target: {IDEON_IP}:{IDEON_PORT}")
    print()
    
    sock = None
    try:
        # 1. Creazione socket
        print("[1] Creazione socket TCP...")
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(TIMEOUT)
        print("    OK - Socket creato")
        
        # 2. Connessione
        print(f"[2] Connessione a {IDEON_IP}:{IDEON_PORT}...")
        sock.connect((IDEON_IP, IDEON_PORT))
        print("    OK - Connesso!")
        
        connected_time = time.strftime("%Y-%m-%d %H:%M:%S")
        print(f"    Timestamp: {connected_time}")
        
        # 3. Invia STATSV (stato macchina)
        print("[3] Invio comando STATSV...")
        statsv_cmd = create_idecon_command("STATSV")
        sock.sendall(statsv_cmd)
        print(f"    TX ({len(statsv_cmd)} bytes): {statsv_cmd.hex().upper()}")
        
        # 4. Ricevi risposta STATSV
        print("[4] Attesa risposta STATSV...")
        response = sock.recv(512)
        if response:
            response_str = parse_response(response)
            print(f"    RX ({len(response)} bytes): {response.hex().upper()}")
            print(f"    Parsed: {response_str}")
        else:
            print("    Nessuna risposta ricevuta")
            
        # 5. Invia STATREQ (richiesta statistiche)
        print()
        print("[5] Invio comando STATREQ (statistiche)...")
        statreq_cmd = create_idecon_command("STATREQ")
        sock.sendall(statreq_cmd)
        print(f"    TX ({len(statreq_cmd)} bytes): {statreq_cmd.hex().upper()}")
        
        # 6. Ricevi risposta STATP
        print("[6] Attesa risposta STATP...")
        response = sock.recv(512)
        if response:
            response_str = parse_response(response)
            print(f"    RX ({len(response)} bytes): {response.hex().upper()}")
            print(f"    Parsed: {response_str}")
        else:
            print("    Nessuna risposta ricevuta")
            
        # 7. Test Keep-alive (attendiamo eventuali pesate automatiche)
        print()
        print("[7] Attesa 2s per eventuali pesate automatiche...")
        time.sleep(2)
        
        # Prova a ricevere dati extra
        sock.setblocking(False)
        try:
            extra_data = sock.recv(512)
            if extra_data:
                print(f"    RX aggiuntivo ({len(extra_data)} bytes): {extra_data.hex().upper()}")
                extra_str = parse_response(extra_data)
                print(f"    Parsed: {extra_str}")
        except BlockingIOError:
            print("    Nessun dato aggiuntivo")
        
        print()
        print("="*60)
        print("TEST COMPLETATO CON SUCCESSO!")
        print("="*60)
        return True
        
    except socket.timeout:
        print(f"\nERRORE: Timeout connessione ({TIMEOUT}s)")
        return False
    except socket.error as e:
        print(f"\nERRORE Socket: {e}")
        return False
    except Exception as e:
        print(f"\nERRORE: {e}")
        return False
    finally:
        if sock:
            sock.close()
            print("\nSocket chiuso")

def quick_test():
    """Test rapido singolo comando"""
    print("QUICK TEST - Invio solo comando STATSV")
    print("-" * 40)
    
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.settimeout(TIMEOUT)
    
    try:
        sock.connect((IDEON_IP, IDEON_PORT))
        print(f"Connesso a {IDEON_IP}:{IDEON_PORT}")
        
        # Invia STATSV
        cmd = b'\x02STATSV\x03'
        sock.sendall(cmd)
        print(f"Inviato: {cmd.hex()}")
        
        # Ricevi
        data = sock.recv(512)
        if data:
            print(f"Ricevuto: {data.hex()}")
            print(f"Risposta: {parse_response(data)}")
        else:
            print("Nessuna risposta")
            
    except Exception as e:
        print(f"Errore: {e}")
    finally:
        sock.close()

if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == '--quick':
        quick_test()
    else:
        test_connection()

File: test_idecon_connection.cs
/*
 * Test Connessione Bilancia IDECON (C#)
 * IP: 172.16.224.210 | Port: 50000
 * Protocollo: STX (0x02) + comando + ETX (0x03)
 */

using System;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace IdeconTest
{
    class Program
    {
        // Configurazione
        private const string IDEON_IP = "172.16.224.210";
        private const int IDEON_PORT = 50000;
        private const int TIMEOUT = 5000; // ms

        static void Main(string[] args)
        {
            Console.WriteLine("==================================================");
            Console.WriteLine("TEST CONNESSIONE BILANCIA IDECON (C#)");
            Console.WriteLine("==================================================");
            Console.WriteLine($"Target: {IDEON_IP}:{IDEON_PORT}");
            Console.WriteLine();

            bool useQuick = args.Length > 0 && args[0] == "--quick";

            if (useQuick)
            {
                QuickTest();
            }
            else
            {
                FullTest();
            }

            Console.WriteLine();
            Console.WriteLine("Test terminato. Premi un tasto per uscire...");
            Console.ReadKey();
        }

        static void FullTest()
        {
            using (TcpClient client = new TdpClient())
            {
                client.ReceiveTimeout = TIMEOUT;
                client.SendTimeout = TIMEOUT;

                try
                {
                    // 1. Connessione
                    Console.WriteLine("[1] Connessione in corso...");
                    client.Connect(IDEON_IP, IDEON_PORT);
                    Console.WriteLine("    OK - Connesso!");

                    NetworkStream stream = client.GetStream();

                    // 2. Invia STATSV
                    Console.WriteLine("[2] Invio comando STATSV...");
                    byte[] statsvCmd = CreateIdeconCommand("STATSV");
                    stream.Write(statsvCmd, 0, statsvCmd.Length);
                    Console.WriteLine($"    TX ({statsvCmd.Length} bytes): {ByteArrayToHex(statsvCmd)}");

                    // 3. Ricevi risposta STATSV
                    Console.WriteLine("[3] Attesa risposta STATSV...");
                    string response = ReadResponse(stream);
                    if (!string.IsNullOrEmpty(response))
                    {
                        Console.WriteLine($"    RX: {response}");
                    }
                    else
                    {
                        Console.WriteLine("    Nessuna risposta ricevuta");
                    }

                    // 4. Invia STATREQ
                    Console.WriteLine();
                    Console.WriteLine("[4] Invio comando STATREQ...");
                    byte[] statreqCmd = CreateIdeconCommand("STATREQ");
                    stream.Write(statreqCmd, 0, statreqCmd.Length);
                    Console.WriteLine($"    TX ({statreqCmd.Length} bytes): {ByteArrayToHex(statreqCmd)}");

                    // 5. Ricevi risposta STATP
                    Console.WriteLine("[5] Attesa risposta STATP...");
                    response = ReadResponse(stream);
                    if (!string.IsNullOrEmpty(response))
                    {
                        Console.WriteLine($"    RX: {response}");
                    }
                    else
                    {
                        Console.WriteLine("    Nessuna risposta ricevuta");
                    }

                    // 6. Test keep-alive (2 secondi)
                    Console.WriteLine();
                    Console.WriteLine("[6] Attesa 2s per eventuali pesate automatiche...");
                    Thread.Sleep(2000);

                    // Verifica dati pendenti
                    if (client.Available > 0)
                    {
                        byte[] extra = new byte[client.Available];
                        stream.Read(extra, 0, extra.Length);
                        Console.WriteLine($"    RX aggiuntivo: {ByteArrayToHex(extra)}");
                    }
                    else
                    {
                        Console.WriteLine("    Nessun dato aggiuntivo");
                    }

                    Console.WriteLine();
                    Console.WriteLine("==================================================");
                    Console.WriteLine("TEST COMPLETATO CON SUCCESSO!");
                    Console.WriteLine("==================================================");
                }
                catch (SocketException ex)
                {
                    Console.WriteLine();
                    Console.WriteLine($"ERRORE Socket [{ex.SocketErrorCode}]: {ex.Message}");
                }
                catch (Exception ex)
                {
                    Console.WriteLine();
                    Console.WriteLine($"ERRORE: {ex.Message}");
                }
            }
        }

        static void QuickTest()
        {
            Console.WriteLine("QUICK TEST - Connessione semplice");
            Console.WriteLine(new string('-', 40));

            using (TcpClient client = new TdpClient())
            {
                client.ReceiveTimeout = TIMEOUT;

                try
                {
                    client.Connect(IDEON_IP, IDEON_PORT);
                    Console.WriteLine($"Connesso a {IDEON_IP}:{IDEON_PORT}");

                    NetworkStream stream = client.GetStream();

                    // Invia STATSV
                    byte[] cmd = CreateIdeconCommand("STATSV");
                    stream.Write(cmd, 0, cmd.Length);
                    Console.WriteLine($"Inviato: {ByteArrayToHex(cmd)}");

                    // Ricevi
                    Thread.Sleep(500);
                    if (client.Available > 0)
                    {
                        byte[] buffer = new byte[512];
                        int bytesRead = stream.Read(buffer, 0, buffer.Length);
                        byte[] response = new byte[bytesRead];
                        Array.Copy(buffer, response, bytesRead);
                        Console.WriteLine($"Ricevuto: {ByteArrayToHex(response)}");
                        Console.WriteLine($"Risposta: {ParseResponse(response)}");
                    }
                    else
                    {
                        Console.WriteLine("Nessuna risposta");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Errore: {ex.Message}");
                }
            }
        }

        static byte[] CreateIdeconCommand(string command)
        {
            // Protocollo: STX (0x02) + comando + ETX (0x03)
            byte[] cmdBytes = Encoding.ASCII.GetBytes(command);
            byte[] result = new byte[cmdBytes.Length + 2];

            result[0] = 0x02; // STX
            Array.Copy(cmdBytes, 0, result, 1, cmdBytes.Length);
            result[result.Length - 1] = 0x03; // ETX

            return result;
        }

        static string ReadResponse(NetworkStream stream)
        {
            byte[] buffer = new byte[512];
            int bytesRead = 0;

            // Leggi con timeout
            int attempts = 0;
            while (attempts < 20 && !stream.DataAvailable && bytesRead == 0)
            {
                Thread.Sleep(100);
                attempts++;
            }

            if (stream.DataAvailable)
            {
                using (MemoryStream ms = new MemoryStream())
                {
                    byte[] temp = new byte[512];
                    int read;

                    // Leggi tutti i dati disponibili
                    while (stream.DataAvailable && (read = stream.Read(temp, 0, temp.Length)) > 0)
                    {
                        ms.Write(temp, 0, read);
                    }

                    return ParseResponse(ms.ToArray());
                }
            }

            return string.Empty;
        }

        static string ParseResponse(byte[] data)
        {
            if (data == null || data.Length == 0)
                return string.Empty;

            // Rimuovi STX (0x02) e ETX (0x03)
            int start = 0;
            int end = data.Length;

            if (data.Length > 0 && data[0] == 0x02)
                start = 1;

            if (data.Length > start && data[data.Length - 1] == 0x03)
                end = data.Length - 1;

            int length = end - start;
            if (length <= 0)
                return string.Empty;

            byte[] payload = new byte[length];
            Array.Copy(data, start, payload, 0, length);

            return Encoding.ASCII.GetString(payload);
        }

        static string ByteArrayToHex(byte[] bytes)
        {
            return BitConverter.ToString(bytes).Replace("-", " ").ToUpper();
        }
    }
}